### PR TITLE
doc: Images section

### PR DESCRIPTION
**Pull request by ru-fu**
Clean up the Images section - part of the content that is in there at the moment doesn't belong there, and a lot of instructional content is missing.

This PR adds this content, partly moving it over from the Getting Started and Advanced Guide on the website, partly adding new content.

### CODE REVIEW

File: doc/howto/images_manage.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+(images-manage-edit)=
+## Edit image properties
+
+To set a specific image property that is located under the `properties` key, enter the following command:
+
+    lxc image set-property <image_ID> <key>
+
+To edit the full image properties, including the top-level properties, enter the following command:
+
+    lxc image edit <image_ID>
```
Comment (ru-fu) [inline doc/howto/images_manage.md:63]: It's kinda awkward that images use the `properties` key for what we call `config` everywhere else, so we need to make a difference between "image properties" and "image properties under the `properties` key".

But I guess there isn't anything we can do about this?

And also, is there any reason for calling it differently for images, or is that just "by accident"?

File: doc/reference/image_format.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+LXD supports a wide variety of compression algorithms for tarballs.
+However, for compatibility purposes, you should use `gzip` or `xz`.
+
+(image-format-unified)=
+### Unified tarball
+
+A unified tarball is a single tarball (usually `*.tar.xz`) that contains the full content of the image, including the metadata, the root file system and optionally the template files.
+
+This is the format that LXD itself produces.
+Therefore, you should use a unified tarball for LXD-specific images.
```
Comment (ru-fu) [inline doc/reference/image_format.md:137]: This comes from the existing documentation. But when I use `lxc image export`, I actually get a split image.
Is the statement wrong? Or does it refer to the images LXD provides on the images server? (But those look like split images to me as well?)

File: doc/howto/images_copy.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+: Assign an alias to the copy of the image.
+
+`--copy-aliases`
+: Copy the aliases that the source image has.
+
+`--auto-update`
+: Keep the copy up-to-date with the original image.
+
+`--vm`
+: Copy an image that can be used to create virtual machines.
```
Comment (ru-fu) [inline doc/howto/images_copy.md:29]: Is that correct?
When launching an instance, I don't need to select a different image depending on whether I want to start a container or a VM. But my guess is that LXD requests a different version of the image under the hood, and that's why I need to be more specific when copying?

**Review by ru-fu — COMMENTED** [review]
Some questions.

File: doc/howto/images_copy.md (hunk 3 above)
Comment (tomponline) [reply doc/howto/images_copy.md:29]: Yes it is correct. When launching an instance you do need to provide the `--vm` flag, which under the hood also passes the equivalent indicator to the image selection process (instance type). So here as we are operating only on images, we do need a `--vm` flag in order to differentiate container images from VM images when referring to them by alias.

File: doc/howto/images_copy.md [hunk 4]
```diff
@@ -6,3 +10,91 @@
...
+
+
+### Direct pushing of the image files
+
+This is mostly useful for air-gapped environments where images cannot be
+directly retrieved from an external server.
+
+In such a scenario, image files can be downloaded on another system using:
+
+    lxc image export ubuntu:22.04
```
Comment (tomponline) [inline doc/howto/images_copy.md]: I think we could make it clearer in this section what the resulting files will be with this command and how they relate to the `import` command below.

When I ran `lxc image export ubuntu:22.04` I got:

- ubuntu-22.04-server-cloudimg-amd64-lxd.tar.xz
- ubuntu-22.04-server-cloudimg-amd64.squashfs

When I ran `lxc image export images:ubuntu/jammy` I got:

- lxd.tar.xz
- rootfs.squashfs

In both scenarios it may not be clear to the end user which file represents `META` and which represents `ROOTFS` in the command below (well maybe in the 2nd example it would be obvious that `rootfs.squashfs` is the ROOTFS, but not in the 1st example).

So perhaps we should explain that the tarball represents the metadata file and the squashfs file contains the rootfs?

Also we could call those `META` and `ROOTFS` placeholders `METADATA_FILE` and `ROOTFS_FILE` perhaps for clarity to indicate that they expect file names there?

File: doc/howto/images_copy.md [hunk 5]
```diff
@@ -6,3 +10,91 @@
...
+local image store with:
+
+    lxc image import META ROOTFS --alias ubuntu-22.04
+
+`lxc image import` supports both unified images (single file) and split
+images (two files) with the example above using the latter.
+
+### File on a remote web server
+
+As an alternative to running a full image server only to distribute a
```
Comment (tomponline) [inline doc/howto/images_copy.md]: In this context what is a "full image server"? Is this a LXD server hosting just images?

If so, we could say something like "a LXD server only to distribute images to users"?

File: doc/howto/images_copy.md (hunk 4 above)
Comment (tomponline) [reply doc/howto/images_copy.md]: Also on the topic of command argument placeholders, in the rest of the docs I think we are using angle brackets to indicate the argument name right? E.g. https://linuxcontainers.org/lxd/docs/master/howto/instances_access_files/#pull-files-from-the-instance-to-the-local-machine

> `lxc file pull <instance_name>/<path_to_file> <local_file_path>`

So I think this should be something like:

```
lxc image import <metadata_file_path> <rootfs_file_path> --alias ubuntu-22.04
```

File: doc/howto/images_copy.md [hunk 6]
```diff
@@ -6,3 +10,91 @@
...
+## Import images
+You can import images, that you:
+
+- built yourself (see [Build Images](#build-images)),
+- downloaded manually (see [Manual Download](#manual-download))
+- exported from images or containers (see [Export Images](#export-images) and [Create Image from Containers](#create-image-from-containers))
+
+#### Import container image
+
+Components:
```
Comment (tomponline) [inline doc/howto/images_copy.md]: I wonder if the components of each image type should be separate section/includes at the top that first define what an image actually is made up of, as this could be useful to the reader WRT to the direct pushing of image files above.

File: doc/howto/images_copy.md [hunk 7]
```diff
@@ -6,3 +10,91 @@
...
+You can also download images manually. For that you need to download the components described [above](#import-images).
+
+#### From official LXD image server
+
+!!! note
+    It is easier to use the usual method with `lxc launch`. Use manual download only if you have a specific reason, like modification of the files before use for example.
+
+**Link to official image server:**
+
+[https://images.linuxcontainers.org/images/](https://images.linuxcontainers.org/images/)
```
Comment (tomponline) [inline doc/howto/images_copy.md]: I think you can just use https://images.linuxcontainers.org here as that gives a nicer branded page with a list of the images available, additionally its what should be used in the `lxc remote` entry.

File: doc/howto/images_copy.md [hunk 8]
```diff
@@ -6,3 +10,91 @@
...
+	lxc image import lxd.tar.xz disk.qcow2 --alias custom-imagename
+
+
+### Manual download
+You can also download images manually. For that you need to download the components described [above](#import-images).
+
+#### From official LXD image server
+
+!!! note
+    It is easier to use the usual method with `lxc launch`. Use manual download only if you have a specific reason, like modification of the files before use for example.
```
Comment (tomponline) [inline doc/howto/images_copy.md]: Maybe rephrase this to something like:

> If you are creating an instance from an image (for example using `lxc launch`) there is no need to manually download the image first as LXD will download it automatically if needed. Only use manual download if you have a specific reason, like modification of the files before use for example.

File: doc/howto/images_create.md [hunk 9]
```diff
@@ -3,3 +3,102 @@
...
+
+### Publishing an instance or snapshot as a new image
+
+An instance or one of its snapshots can be turned into a new image.
+This is done on the CLI with `lxc publish`.
+
+When doing this, you will most likely first want to cleanup metadata and
+templates on the instance you're publishing using the `lxc config metadata`
+and `lxc config template` commands. You will also want to remove any
+instance-specific state like host SSH keys, `dbus/systemd machine-id`, ...
```
Comment (tomponline) [inline doc/howto/images_create.md]: I think its important to say "inside the instance" here to clarify these are not LXD settings, e.g.

> You will also want to remove any
instance-specific state inside the instance, such as host SSH keys, `dbus/systemd machine-id`, etc.

File: doc/howto/images_create.md [hunk 10]
```diff
@@ -3,3 +3,102 @@
...
+| `targets` | defines configs for specific targets (e.g. LXD-client, instances etc.) |  see [targets.md](https://github.com/lxc/distrobuilder/blob/master/doc/targets.md) |
+| `files` | defines generators to modify files | see [generators.md](https://github.com/lxc/distrobuilder/blob/master/doc/generators.md) |
+| `packages` | defines packages for install or removal; add repositories |   see [packages.md](https://github.com/lxc/distrobuilder/blob/master/doc/packages.md) |
+| `actions` | defines scripts to be run after specific steps during image building |  see [actions.md](https://github.com/lxc/distrobuilder/blob/master/doc/actions.md) |
+| `mappings` | maps different terms for architectures for specific distributions (e.g. x86_64: amd64) | see [mappings.md](https://github.com/lxc/distrobuilder/blob/master/doc/mappings.md) |
+
+
+!!! note "Note for VMs"
+	You should either build an image with cloud-init support (provides automatic size growth) or set a higher size in the template, because the standard size is relatively small (~4 GB). Alternatively you can also grow it manually.
```
Comment (tomponline) [inline doc/howto/images_create.md]: > provides automatic root disk filesystem size growing inside the guest

File: doc/image-handling.md [hunk 11]
```diff
@@ -1,82 +1,63 @@
...
-on the target LXD.
+Each image is identified by a fingerprint (SHA256).
+To make it easier to manage images, LXD allows defining one or more aliases for each image.
 
-## Sources
+## Caching
 
-LXD supports importing images from three different sources:
+When you create an instance using a remote image, LXD downloads the image and caches it locally.
+It is stored in the local image store with the cached bit set.
```
Comment (tomponline) [inline doc/image-handling.md:21]: Maybe use "indicator" or "flag" here rather than "bit", as that sounds a bit low level/technical to me (its not actually using a bit anyway ;)).

File: doc/howto/images_remote.md [hunk 12]
```diff
@@ -1,125 +1,73 @@
...
 # How to use remote images
 
-- Link to {ref}`remote-image-servers`
-- List remotes
-- Add remotes
-- Set a default remote
-- Reference an image (`<remote>:<image>`)
-- Search for images
-- Configure caching
+LXD is pre-configured with several remote image servers.
```
Comment (tomponline) [inline doc/howto/images_remote.md:4]: Technically speaking the `lxc` CLI command is preconfigured with several remote image servers.
LXD itself has no concept of that.

Maybe we can say this here for accuracy, something like:

> The LXC CLI command is...

File: doc/howto/images_remote.md [hunk 13]
```diff
@@ -1,125 +1,73 @@
...
+    images:ubuntu/22.04
+    ubuntu:22.04
+    local:ed7509d7e83f
 
-Show all 64-bit Debian images:
+(images-remote-default)=
+## Select a default remote
 
-	lxc image list images: debian amd64
+If you specify an image name without the name of the remote, the default image server is used.
```
Comment (tomponline) [inline doc/howto/images_remote.md:65]: Should we say which server is the default?

Comment (tomponline) [reply doc/howto/images_remote.md:65]: Ah I see you covered that below.

File: doc/howto/instances_create.md [hunk 14]
```diff
@@ -60,6 +60,11 @@ To launch a virtual machine with a Ubuntu 22.04 image from the `images` server u
 
     lxc launch images:ubuntu/22.04 ubuntu-vm --vm
 
+```{tip}
+When creating a virtual machine, use the `cloud` variant of an image (you can identify it by checking for the `cloud`-tag in the image alias).
```
Comment (tomponline) [inline doc/howto/instances_create.md:64]: Why is this tip specific to VMs? Also this feels like it should be a separate commit.

The cloud variants are available for both containers and VMs. 
One thing that might be worth pointing out (if not already done so) is that using the cloud-init variant will allow for root filesystem resizing to the configured root disk size, whereas otherwise it needs to be done manually.

Comment (tomponline) [reply doc/howto/instances_create.md:64]: Also the `cloud` variant is only really relevant to the `images:` remote.

File: doc/reference/remote_image_servers.md [hunk 15]
```diff
@@ -1,8 +1,43 @@
 (remote-image-servers)=
 # Remote image servers
 
-LXD comes with 3 default servers:
+LXD comes pre-configured with the following default remote image servers:
```
Comment (tomponline) [inline doc/reference/remote_image_servers.md:4]: > The `lxc` CLI command comes....

File: doc/reference/remote_image_servers.md [hunk 16]
```diff
@@ -1,8 +1,43 @@
...
+## Remote server types
+
+LXD supports the following types of remote image servers:
+
+Simple streams servers
+: Pure image servers that use the [simple streams format](https://git.launchpad.net/simplestreams/tree/).
+  The default image servers are simple streams servers.
+
+Public LXD servers
+: Empty LXD servers with no storage pools and no networks that serve solely as image servers.
```
Comment (tomponline) [inline doc/reference/remote_image_servers.md:34]: You might have a storage pool if you want to store the images in a custom volume.
Perhaps just say here:

> LXD servers that are used solely to serve images (they do not run instances themselves).

File: doc/howto/images_manage.md [hunk 17]
```diff
@@ -1,20 +1,105 @@
...
-## Export images
-Use:
+    lxc image list [<remote>:]
 
-	lxc image export imagename [target folder] [flags]
+If you do not specify a remote, the {ref}`default remote <images-remote-default>` is used.
 
-Flags:
+(images-manage-filter)=
+### Filter available images
```
Comment (tomponline) [inline doc/howto/images_manage.md:16]: I did not know this, thanks :)

File: doc/howto/images_manage.md (hunk 1 above)
Comment (tomponline) [reply doc/howto/images_manage.md:63]: One for @stgraber as I don't know the history on this.

File: doc/howto/images_manage.md [hunk 18]
```diff
@@ -1,20 +1,105 @@
...
+
+You can see some of the existing aliases in the image list.
+To see the full list, enter the following command:
+
+    lxc image alias list
+
+You can directly assign an alias to an image when you {ref}`copy or import <images-copy>` or {ref}`publish <images-create-publish>` it.
+Alternatively, enter the following command:
+
+    lxc image alias create <alias> <image_fingerprint>
```
Comment (tomponline) [inline doc/howto/images_manage.md:78]: We should probably use `<alias_name>` here to align with other examples, such as https://linuxcontainers.org/lxd/docs/master/howto/instances_create/#usage where it uses `<image_name> <instance_name>`.

File: doc/howto/images_manage.md [hunk 19]
```diff
@@ -1,20 +1,105 @@
...
+(images-manage-export)=
+## Export an image to a file
+
+Images are located in the image store of your local server or a remote LXD server.
+You can export them to a file though.
+This method can be useful to back up image files or to transfer them to an air-gapped environment.
+
+To export a container image to a file, enter the following command:
+
+    lxc image export [<remote>:]<image> [<output_location>]
```
Comment (tomponline) [inline doc/howto/images_manage.md:99]: Maybe `<output_directory_path>` here? This would help to indicate that it requires a directory path (as opposed to an image file name, which one may expect when exporting an image).

File: doc/howto/images_create.md [hunk 20]
```diff
@@ -1,104 +1,40 @@
...
 
-	lxc publish
+Before you publish an image from an instance, clean up all data that should not be included in the image.
+Usually, this includes the following data:
 
-## Build images
-For building your own images, you can use [`distrobuilder`](https://github.com/lxc/distrobuilder) (a tool developed by us).
+- Instance metadata (use `lxc config metadata` to edit)
+- File templates (use `lxc config template` to edit)
+- Instance-specific data (for example, host SSH keys and `dbus/systemd machine-id`)
```
Comment (tomponline) [inline doc/howto/images_create.md:33]: How about:

> Instance-specific data inside the instance itself

File: doc/howto/images_copy.md [hunk 21]
```diff
@@ -1,100 +1,88 @@
...
 
-#### Import container image
+    lxc image import <image_file_or_dir> [<target_remote>:]
 
-Components:
+To import a split image, enter the following command:
 
-- lxd.tar.xz
-- rootfs.squashfs
+    lxc image import <metadata_tarball> <rootfs_tarball> [<target_remote>:]
```
Comment (tomponline) [inline doc/howto/images_copy.md:52]: Can you add `_path` to these to avoid any confusion around what these arguments are (i.e they are not the file contents themselves).

File: doc/howto/images_copy.md [hunk 22]
```diff
@@ -1,100 +1,88 @@
...
+To import an image from the local file system, use the `lxc image import` command.
+This command supports both {ref}`unified images <image-format-unified>` (compressed file or directory) and {ref}`split images <image-format-split>` (two files).
 
-- built yourself (see [Build Images](#build-images)),
-- downloaded manually (see [Manual Download](#manual-download))
-- exported from images or containers (see [Export Images](#export-images) and [Create Image from Containers](#create-image-from-containers))
+To import a unified image from one file or directory, enter the following command:
 
-#### Import container image
+    lxc image import <image_file_or_dir> [<target_remote>:]
```
Comment (tomponline) [inline doc/howto/images_copy.md:48]: Maybe `image_file_or_directory_path`? Bit long perhaps.

File: doc/howto/images_copy.md [hunk 23]
```diff
@@ -1,100 +1,88 @@
...
-- lxd.tar.xz
-- disk.qcow2
+    lxc image import <URL>
 
-Use:
+You can assign an alias to the local image with the `--alias` flag.
 
-	lxc image import lxd.tar.xz disk.qcow2 --alias custom-imagename
+(images-copy-http-headers)=
+#### Custom HTTP headers
```
Comment (tomponline) [inline doc/howto/images_copy.md:72]: This is nice, I didn't know about this feature :)

File: doc/reference/image_format.md [hunk 24]
```diff
@@ -1,72 +1,75 @@
...
-In this mode the image identifier is the SHA-256 of the concatenation of
-the metadata and rootfs tarball (in that order).
+### Root file system
 
-### Supported compression
+For containers, the `rootfs/` directory contains a full file system tree of the root directory (`/`) in the container.
 
-LXD supports a wide variety of compression algorithms for tarballs
-though for compatibility purposes, `gzip` or `xz` should be preferred.
+Virtual machines use a `rootfs.img` file instead of a `rootfs/` directory.
```
Comment (tomponline) [inline doc/reference/image_format.md:56]: Can you say "Virtual machines use a `rootfs.img` qcow2 file instead..." 
I know you explained it further down but would be useful to briefly mention it here too.

File: doc/reference/image_format.md (hunk 2 above)
Comment (tomponline) [reply doc/reference/image_format.md:137]: I think this is referring to the format of the image produced when using `lxc publish`, as doing that and then using `lxc image export` downloads a single tarball file.

So `lxc image export` seems to only be exporting the image in its current format (whatever that was from the remote server or locally published) and not "producing" it. It is unclear though I agree.

**Review by tomponline — COMMENTED** [review]
Apologies for some of the comments on the "moving" commit. Thanks for highlighting that.

The new content is great, there's just a few nits about particular wording, but apart from that good to go.

File: doc/howto/instances_create.md (hunk 14 above)
Comment (ru-fu) [reply doc/howto/instances_create.md:64]: This note comes from the Getting Started guide: https://linuxcontainers.org/lxd/getting-started-cli/#images-for-virtual-machines
I'm not sure why it was added there, but it was part of the images section - I don't feel that is where it belongs though, that's why I moved it to the instance creation.

And my assumption was that using the cloud images is especially important for VMs, because those need the lxd-agent to be able to be able to log on - but that might be wrong. ;)

So do we generally want to recommend using the cloud variants?

File: doc/reference/image_format.md (hunk 2 above)
Comment (ru-fu) [reply doc/reference/image_format.md:137]: OK, reworded a bit - and I removed the recommendation since we're apparently not following it ourselves. ;)

**Comment by ru-fu** [comment]
Thanks a lot for the comments! Updated now.

A few questions are still open though.

File: doc/reference/image_format.md (hunk 2 above)
Comment (tomponline) [reply doc/reference/image_format.md]: @stgraber what do you think about this?

File: doc/howto/instances_create.md (hunk 14 above)
Comment (tomponline) [reply doc/howto/instances_create.md:64]: The `lxd-agent` is included in all VM images we can get it into, so its not specific to having a cloud variant.
Also the concept of a "cloud variant" is only relevant for the "images:" remote and not for the "ubuntu:" remote (as these always include cloud-init and there is no "cloud variant").

I agree it makes sense to keep it in instance creation section.

I agree its usually a good idea to use a VM variant that has cloud-init in it so as to get the root filesystem auto-grow feature.
However I think we should explain a bit more about which images contain cloud-init and why that is recommended for VMs particularly (for the reason above).

Comment (tomponline) [reply doc/howto/instances_create.md:64]: Also having an example of an image address with a "cloud tag" would be good as its not so clear here currently.

Comment (ru-fu) [reply doc/howto/instances_create.md:64]: Actually, I'm starting to think we should probably leave this change out for now.
From what I understand, we should include the information about the root file system auto-grow feature in the cloud-init docs when we clean them up. And then we can add a hint here about what image to select for a VM and why. But including the full explanation here would be out of place imo.
Does that sound okay? Then I'll make a note of it for cloud-init and remove this change from the PR.

Comment (tomponline) [reply doc/howto/instances_create.md:64]: Sounds good.

Comment (ru-fu) [reply doc/howto/instances_create.md]: OK, updated. :)

**Comment by ru-fu** [comment]
Updated again - that leaves only two open questions for @stgraber :)

File: doc/howto/images_manage.md (hunk 1 above)
Comment (stgraber) [reply doc/howto/images_manage.md:63]: It's not an accident, though indeed the naming of `properties` is a bit unfortunate.

We very specifically do not want images to be able to provide configuration to LXD as this would turn into a very easy attack vector. To make this very clear to users and to avoid us accidentally introducing actual configuration as part of what's shipped by an image, we made sure not to have a `config` on images, but just `properties` which don't alter LXD's behavior and is effectively purely cosmetic.

File: doc/reference/image_format.md (hunk 2 above)
Comment (stgraber) [reply doc/reference/image_format.md]: `This is the format that LXD itself produces.` is indeed still accurate.
LXD does not produce the images it downloads from the various image servers, those are produced by tools like distrobuilder. LXD produces images when one uses `lxc publish` and when one does that, the result is indeed a unified image.

The recommendation still mostly stands that if you're going to be building images by hand for consumption by LXD alone, then you should go with the unified format as it's far easier to understand and far harder to mess up.

Using the split format is very convenient for cases where images may be consumed by other tools than LXD.
That's why the official Ubuntu images are in that format (consumed by Docker, LXD, ...) as well as why our own images are also in that format (consumed by both LXC and LXD).

File: doc/howto/images_manage.md (hunk 1 above)
Comment (ru-fu) [reply doc/howto/images_manage.md:63]: Thanks! I added a note to clarify that: "These properties can be used to convey information about the image. They do not configure LXD’s behavior in any way."

File: doc/reference/image_format.md (hunk 2 above)
Comment (ru-fu) [reply doc/reference/image_format.md]: Updated as follows for the unified tarballs:
"This is the format that LXD itself uses internally when publishing images. It is usually easier to work with; therefore, you should use the unified format when creating LXD-specific images."

And for the split tarballs:
"This format is designed to allow for easy image building from existing non-LXD rootfs tarballs that are already available. You should also use this format if you want to create images that can be consumed by both LXD and other tools."

File: doc/howto/images_manage.md (hunk 1 above)
Comment (stgraber) [reply doc/howto/images_manage.md]: Sounds good!

File: doc/reference/image_format.md (hunk 2 above)
Comment (stgraber) [reply doc/reference/image_format.md]: Sounds good!
